### PR TITLE
sql: add EXPLAIN (VEC) output to the stmt bundle

### LINKED ISSUE #47940: sql: put the output of EXPLAIN (VEC) in the EXPLAIN ANALYZE (DEBUG) bundle
EXPLAIN ANALYZE (DEBUG) generates a bundle which contains the output of most flavors of EXPLAIN, except EXPLAIN (VEC). I tried to add it but it looks very hard to integrate with. For example, the code messes with the subqueries in `planner.curPlan`:
https://github.com/cockroachdb/cockroach/blob/a1296f903446c6ca17806bcb288d2bf17c85570e/pkg/sql/explain_plan.go#L216
https://github.com/cockroachdb/cockroach/blob/a1296f903446c6ca17806bcb288d2bf17c85570e/pkg/sql/explain_vec.go#L59

I believe this code could use some cleanup. Also note that `EXPLAIN (VEC)` doesn't work when there are subqueries. But in the case of `EXPLAIN ANALYZE (DEBUG)` we actually run everything. So wecould conceivably generate it, perhaps separately for each subquery/postquery. This is what we do with the distsql diagrams: the bundle contains a separate diagram for each subquery/main query/postquery.

### PR DESCRIPTION
**Pull request by yuzefovich**
**sql: delay saving of flows till after the setup**

This commit reorders the code that deals with the flow specifications
a bit. Previously, we would call `saveFlows` right after generating flow
specifications and we would release the resources back to their pools
once the setup request for each flow has been sent out. However, the
follow-up commit will include the generation of `EXPLAIN (VEC, VERBOSE)`
diagram into the statement bundle, and to get there we need to delay
both of those actions until we know that the flows are vectorized. As
a result, now `saveFlows` is called after `setupFlows` (when we know
whether they were vectorized or not) and the release of the flow specs
is deferred.

Release note: None

**sql: add EXPLAIN (VEC) output to the stmt bundle**

This commit refactors the code of `EXPLAIN (VEC)` to be reused to add
`(VEC)` and `(VEC, VERBOSE)` flavours to the statement bundle. The code
to print the diagram is moved into `colflow` package, but other changes
are minor.

The addition of these diagrams is performed on the best effort basis
(meaning that if an error is encountered during populating the diagram,
the error is ignored) because in some edge cases the errors are expected
to occur.

Fixes: #47940.

Release note (sql change): Statement diagnostics bundles now contain
output of `EXPLAIN (VEC)` and `EXPLAIN (VEC, VERBOSE)` commands for the
statements.

### CODE REVIEW

**Comment by yuzefovich** [comment]
Confirmed manually that it works:
Screen Shot 2021-03-11 at 6 06 08 PM

**Review by RaduBerinde — COMMENTED** [review]
Very cool! :lgtm: 

The lists of files in `TestExplainAnalyzeDebug` need updating.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61887#-:-MV_BmWvESp3JURxK8WX:b8sykt0)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[pkg/sql/distsql_physical_planner.go, line 700 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/61887#-MV_By_O-44zRIyJfO1W:-MV_By_O-44zRIyJfO1X:bmvwc4o) ([raw file](https://github.com/cockroachdb/cockroach/blob/186d34931db5e42dcb1b8de36d7798b45499a28d/pkg/sql/distsql_physical_planner.go#L700)):*
> ```Go
> 		var explainVecVerbose []string
> 		if planner.instrumentation.collectBundle && planner.curPlan.flags.IsSet(planFlagVectorized) {
> 			flowCtx := newFlowCtxForExplainPurposes(p, planner, &planner.extendedEvalCtx.DistSQLPlanner.rpcCtx.ClusterID)
> ```

Given that a panic in this code could be very bad, consider adding a panic catcher in this function (you can copy the block in `emitExplain()`).
___
*[pkg/sql/explain_bundle.go, line 324 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/61887#-MV_QqiCAkzRpbul0qsb:-MV_QqiD0MfKOt5mM5nK:b-ik7rol) ([raw file](https://github.com/cockroachdb/cockroach/blob/186d34931db5e42dcb1b8de36d7798b45499a28d/pkg/sql/explain_bundle.go#L324)):*
> ```Go
> 			}
> 			if len(d.explainVec) > 0 {
> 				b.z.AddFile(fmt.Sprintf(filenameFormat, ""), strings.Join(d.explainVec, "\n"))
> ```

[nit] The filenameFormat logic is hard to follow (especially since we're intermixing it with other Sprintfs). I'd do something simpler:
```
extra := ""
if len() > 1 {
  extra = fmt.Sprintf("-%d-%s")
}
AddFile(fmt.Sprintf("vec%s.txt", extra))
AddFile(fmt.Sprintf("vec%s-v.txt", extra))
```

**Review by yuzefovich — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61887)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @RaduBerinde)

___
*[pkg/sql/distsql_physical_planner.go, line 700 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/61887#-MV_By_O-44zRIyJfO1W:-MVbhBgG9JAydWqIbyZz:b-unwseh) ([raw file](https://github.com/cockroachdb/cockroach/blob/186d34931db5e42dcb1b8de36d7798b45499a28d/pkg/sql/distsql_physical_planner.go#L700)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Given that a panic in this code could be very bad, consider adding a panic catcher in this function (you can copy the block in `emitExplain()`).
</blockquote></details>

We already had the vectorized panic catcher in `ExplainVec`, and I extended it to the whole function body instead of introducing a panic catcher here.

**Comment by yuzefovich** [comment]
TFTR!

bors r+

**Comment by yuzefovich** [comment]
Thoughts on backporting this to 21.1? I think it should be pretty safe and somewhat beneficial.

**Comment by jordanlewis** [comment]
I think this is a good idea to backport.
